packages/firestore/test/unit/model/values.test.ts

```diff
@@ -0,0 +1,420 @@
+ */
+
+import * as api from '../../../src/protos/firestore_proto_api';
+
+import { expect } from 'chai';
+
+import { Timestamp } from '../../../src/api/timestamp';
+import { GeoPoint } from '../../../src/api/geo_point';
+import { DatabaseId } from '../../../src/core/database_info';
+import {
+  canonicalId,
+  compare,
+  equals,
+  estimateByteSize
+} from '../../../src/model/values';
+import { DocumentKey } from '../../../src/model/document_key';
+import { PrimitiveValue } from '../../../src/model/proto_field_value';
+import { primitiveComparator } from '../../../src/util/misc';
+import {
+  blob,
+  dbId,
+  expectCorrectComparisonGroups,
+  expectEqualitySets,
+  key,
+  ref
+} from '../../util/helpers';
+import { valueOf } from '../../util/values';
+
+describe('Values', () => {
+  const date1 = new Date(2016, 4, 2, 1, 5);
+  const date2 = new Date(2016, 5, 20, 10, 20, 30);
+
+  it('compares values for equality', () => {
+    // Each subarray compares equal to each other and false to every other value
+    const values: api.Value[][] = [
+      [wrap(true), wrap(true)],
+      [wrap(false), wrap(false)],
+      [wrap(null), wrap(null)],
+      [wrap(0 / 0), wrap(Number.NaN), wrap(NaN)],
+      // -0.0 and 0.0 order the same but are not considered equal.
+      [wrap(-0.0)],
+      [wrap(0.0)],
+      [wrap(1), { integerValue: 1 }],
+      // Doubles and Integers order the same but are not considered equal.
+      [{ doubleValue: 1.0 }],
+      [wrap(1.1), wrap(1.1)],
+      [wrap(blob(0, 1, 2)), wrap(blob(0, 1, 2))],
+      [wrap(blob(0, 1))],
+      [wrap('string'), wrap('string')],
+      [wrap('strin')],
+      // latin small letter e + combining acute accent
+      [wrap('e\u0301b')],
+      // latin small letter e with acute accent
+      [wrap('\u00e9a')],
+      [wrap(date1), wrap(Timestamp.fromDate(date1))],
+      [wrap(date2)],
+      // TODO(mrschmidt): Support ServerTimestamps
+      // [
+      //   // NOTE: ServerTimestampValues can't be parsed via wrap().
+      //   serverTimestamp(Timestamp.fromDate(date1), null),
+      //   serverTimestamp(Timestamp.fromDate(date1), null)
+      // ],
+      // [serverTimestamp(Timestamp.fromDate(date2), null)],
+      [wrap(new GeoPoint(0, 1)), wrap(new GeoPoint(0, 1))],
+      [wrap(new GeoPoint(1, 0))],
+      [wrap(ref('project', 'coll/doc1')), wrap(ref('project', 'coll/doc1'))],
+      [wrap(ref('project', 'coll/doc2'))],
+      [wrap(['foo', 'bar']), wrap(['foo', 'bar'])],
+      [wrap(['foo', 'bar', 'baz'])],
+      [wrap(['foo'])],
+      [wrap({ bar: 1, foo: 2 }), wrap({ foo: 2, bar: 1 })],
+      [wrap({ bar: 2, foo: 1 })],
+      [wrap({ bar: 1, foo: 1 })],
+      [wrap({ foo: 1 })]
+    ];
+    expectEqualitySets(values, (v1, v2) => equals(v1, v2));
+  });
+
+  it('normalizes values for equality', () => {
+    // Each subarray compares equal to each other and false to every other value
+    const values: api.Value[][] = [
+      [{ integerValue: '1' }, { integerValue: 1 }],
```

Comment:
Isn't there also a case where there's a nested value object?

Reply:
If you are referring to the `Int32Proto`, this seems to only be used in a target's `limit` specification, which allows us to encode a difference between `0` and unset. I don't see any usage for this patten in api.Value.

Reply:
Ah, I see. Right.